op-chain-ops/cmd/withdrawals/main.go

```diff
@@ -33,60 +165,518 @@ func main() {
+					if err != nil {
+						return err
+					}
+					initialTime := block.Time()
+					for {
+						log.Info("waiting for finalization")
+						if block.Time() >= initialTime+period.Uint64() {
+							log.Info("can be finalized")
+							break
+						}
+						time.Sleep(1 * time.Second)
+						block, err = l1Client.BlockByNumber(context.Background(), nil)
+						if err != nil {
+							return err
+						}
+					}
+				} else {
+					log.Info("Withdrawal already proven to OptimismPortal")
+				}
+
+				// check to see if the withdrawal has been finalized already
+				isFinalized, err := portal.FinalizedWithdrawals(&bind.CallOpts{}, hash)
+				if err != nil {
+					return err
+				}
+
+				if !isFinalized {
+					log.Info("Finalizing withdrawal")
+
+					// Get the ETH balance of the withdrawal target *before* the finalization
+					targetBalBefore, err := l1Client.BalanceAt(context.Background(), common.BytesToAddress(wd.Target.Bytes()), nil)
+					if err != nil {
+						return err
+					}
+
+					log.Debug(fmt.Sprintf("Target balance before finalization: %v", targetBalBefore))
+
+					// Finalize withdrawal
+					tx, err := portal.FinalizeWithdrawalTransaction(
+						opts,
+						wdTx,
+					)
+					if err != nil {
+						return err
+					}
+
+					receipt, err := bind.WaitMined(context.Background(), l1Client, tx)
+					if err != nil {
+						return err
+					}
+					if receipt.Status != types.ReceiptStatusSuccessful {
+						return errors.New("withdrawal finalize unsuccessful")
+					}
+
+					log.Info("withdrawal finalized", "tx-hash", tx.Hash(), "withdrawal-hash", hash)
+
+					// fetch the call trace
+					var finalizationTrace callFrame
+					tracer := "callTracer"
+					traceConfig := tracers.TraceConfig{
+						Tracer: &tracer,
+					}
+					err = l1RpcClient.Call(&finalizationTrace, "debug_traceTransaction", receipt.TxHash, traceConfig)
+					if err != nil {
+						return err
+					}
+
+					callFrame := findWithdrawalCall(&finalizationTrace, wd, l1xdmAddr)
+					if callFrame == nil {
+						return errors.New("cannot find callframe")
+					}
+
+					traceJson, err := json.MarshalIndent(callFrame, "", "    ")
+					if err != nil {
+						return err
+					}
+					log.Info(fmt.Sprintf("%v", string(traceJson)))
+
+					erc20Abi, err := bindings.ERC20MetaData.GetAbi()
+					if err != nil {
+						return err
+					}
+					transferEvent := erc20Abi.Events["Transfer"]
+
+					abi, err := bindings.L1StandardBridgeMetaData.GetAbi()
+					if err != nil {
+						return err
+					}
+
+					calldata := hexutil.MustDecode(callFrame.Input)
+
+					// this must be the L1 standard bridge
+					method, err := abi.MethodById(calldata)
+					// Handle L1StandardBridge specific logic
+					if err == nil {
+						args, err := method.Inputs.Unpack(calldata[4:])
+						if err != nil {
+							return err
+						}
+
+						log.Info("decoded calldata", "name", method.Name)
+
+						switch method.Name {
```

Comment:
This switch / these case statements are good candidates to pull out into their own function.